lib/services/unified_resource.go

```diff
@@ -0,0 +1,514 @@
+
+	"github.com/google/btree"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport"
+	apidefaults "github.com/gravitational/teleport/api/defaults"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// UnifiedResourceCacheConfig is used to configure a UnifiedResourceCache
+type UnifiedResourceCacheConfig struct {
+	// BTreeDegree is a degree of B-Tree, 2 for example, will create a
+	// 2-3-4 tree (each node contains 1-3 items and 2-4 children).
+	BTreeDegree int
+	// Clock is a clock for time-related operations
+	Clock clockwork.Clock
+	// Component is a logging component
+	Component string
+	ResourceWatcherConfig
+	ResourceGetter
+}
+
+// UnifiedResourceCache contains a representation of all resources that are displayable in the UI
+type UnifiedResourceCache struct {
+	mu  sync.Mutex
+	log *log.Entry
+	cfg UnifiedResourceCacheConfig
+	// tree is a BTree with items
+	tree            *btree.BTreeG[*item]
+	initializationC chan struct{}
+	stale           bool
+	once            sync.Once
+	cache           *utils.FnCache
+	ResourceGetter
+}
+
+// NewUnifiedResourceCache creates a new memory cache that holds the unified resources
+func NewUnifiedResourceCache(ctx context.Context, cfg UnifiedResourceCacheConfig) (*UnifiedResourceCache, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err, "setting defaults for unified resource cache")
+	}
+
+	lazyCache, err := utils.NewFnCache(utils.FnCacheConfig{
+		Context: ctx,
+		TTL:     15 * time.Second,
+		Clock:   cfg.Clock,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	m := &UnifiedResourceCache{
+		log: log.WithFields(log.Fields{
+			trace.Component: cfg.Component,
+		}),
+		cfg: cfg,
+		tree: btree.NewG(cfg.BTreeDegree, func(a, b *item) bool {
+			return a.Less(b)
+		}),
+		initializationC: make(chan struct{}),
+		ResourceGetter:  cfg.ResourceGetter,
+		cache:           lazyCache,
+		stale:           true,
+	}
+
+	if err := newWatcher(ctx, m, cfg.ResourceWatcherConfig); err != nil {
+		return nil, trace.Wrap(err, "creating unified resource watcher")
+	}
+	return m, nil
+}
+
+// CheckAndSetDefaults checks and sets default values
+func (cfg *UnifiedResourceCacheConfig) CheckAndSetDefaults() error {
+	if cfg.BTreeDegree <= 0 {
+		cfg.BTreeDegree = 8
+	}
+	if cfg.Clock == nil {
+		cfg.Clock = clockwork.NewRealClock()
+	}
+	if cfg.Component == "" {
+		cfg.Component = teleport.ComponentUnifiedResource
+	}
+	return nil
+}
+
+// put stores the value into backend (creates if it does not
+// exist, updates it otherwise)
+func (c *UnifiedResourceCache) put(ctx context.Context, i item) error {
+	if len(i.Key) == 0 {
+		return trace.BadParameter("missing parameter key")
+	}
+	return c.read(ctx, func(tree *btree.BTreeG[*item]) error {
+		tree.ReplaceOrInsert(&i)
+		return nil
+	})
```

Comment:
Insertion is always supposed to hit the primary tree.

```suggestion
	if len(i.Key) == 0 {
		return trace.BadParameter("missing parameter key")
	}
	c.mu.Lock()
	defer c.mu.Unlock()
	c.tree.ReplaceOrInsert(&i)
	return nil
```